odoo/fields.py

```diff
@@ -2832,7 +2832,7 @@ def convert_to_record(self, value, record):
             'active' in corecords
             and self.context.get('active_test', record.env.context.get('active_test', True))
         ):
-            corecords = corecords.filtered('active').with_prefetch(prefetch_ids)
+            corecords = corecords.search([("id", "in", corecords._ids), ("active", "=", True)])
```

Comment:
The method was clearly not designed to handle pathological cases like yours.  The whole point of that line was to filter out inactive records by relying on cache _instead of calling search_.  On smaller values, it is much faster.  Your patch clearly improves pathological cases but at the cost of making smaller cases slower.  I don't think it is a good idea.

I suggest instead to avoid going through that line.  Do you actually need your x2many field to contain inactive records?  If not, you can parameterize it to contain active records only (make it explicit in the domain) and not depend on the context for that (add `active_test=False` in the context):
```py
variant_ids = fields.Many2many(..., domain=[('active', '=', True)], context={'active_test': False})
```
In that case, you will never call `filtered()` on it.